Code for Mail Source of App starter project

aamol opened the pull request:
Reference to spring-cloud/spring-cloud-stream-modules#255 .

I have implemented mail-source which I wanted to use as part of migration from Spring-XD to Spring-Could-DataFlow.

Raising pull request if code suits fine. Please feel free to provide suggestion/improvements

Thanks,
Amol

Commit:
Code and test for Mail Source of App starter project

sobychacko commented:
@aamol Thanks for the contribution. Couple of things to note.

- Add this new source in the app-generator project. Please see here - https://github.com/spring-cloud/spring-cloud-stream-app-starters/blob/master/spring-cloud-stream-app-generator/pom.xml
- No need to provide spring.factories as the Configuration class will be imported into the apps generated.
- Please update the descriptor here: https://github.com/spring-cloud/spring-cloud-stream-app-starters/tree/master/spring-cloud-stream-app-descriptor

@garyrussell @artembilan  - Any further thoughts/comments?

artembilan commented:
@sobychacko , it is on my radar already.
Will review a bit later today.

(That will be my first exercise with `SCSt-apps` :wink: ) 

Thank you for heads-up anyway! :smile: 

And yes. Thank you for contribution @aamol !

Commit:
encorporated comments from @sobychacko

artembilan commented on app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mail/PoorMansMailServer.java:
```diff
@@ -0,0 +1,484 @@
+ * @author Gary Russell
+ *
+ */
+public class PoorMansMailServer {
```
@garyrussell , I wonder if we can just move this useful class into the `SI-test` module and don't have such a copies everywhere...

garyrussell replied:
Yes, good idea.

garyrussell replied:
Yes, good idea, but we should probably give it a more sensible name like BasicTestMailServer.

artembilan commented on mail/pom.xml:
```diff
@@ -0,0 +1,18 @@
+	<parent>
+		<groupId>org.springframework.cloud.stream.app</groupId>
+		<artifactId>spring-cloud-stream-app-starters</artifactId>
+		<version>1.0.0.BUILD-SNAPSHOT</version>
```
I'm new to this project and I'm not so familiar (don't like ?) with Maven, so I might will ask a lot of noisy questions which may be obvious or know somehow for the team around.
So, my apologies if I will be so noobie :smile: 

Well, I see that this implicit version is used everywhere in each application.
Can't we extract this value somehow in the external resource and don't have a hell in the future to jump around to fix this value after each release? 
Thanks

sobychacko replied:
@artembilan we don't have to manually change the version value after each release. Bamboo automatically changes that during a CI promotion build. (For ex M1 -> BUILD-SNAPSHOT).

artembilan replied:
Throughout all poms? Than OK.
Thank you anyway for feedback. I was going to think that I play squash already :wink:

artembilan commented on mail/spring-cloud-starter-stream-source-mail/pom.xml:
```diff
@@ -0,0 +1,65 @@
+		<dependency>
+			<groupId>org.springframework.cloud.stream.app</groupId>
+			<artifactId>app-starters-test-support</artifactId>
+		</dependency>
```
Can't we move all the testing artifacts to this `app-starters-test-support` module?
I even see that `spring-integration-test` is already there.
And would be better to reorder dependencies to make it clear what is `production` and what is `test`.
At a glance I was going to comment that we need `<scope>test</scope>` for these artifacts.
And the real picture I can see only after jumping into the `dependencyManagement`...

From other side need to revise if we really need all these dependencies for testing.

aamol replied:
Grouped test dependencies in a section at the last

artembilan commented on mail/spring-cloud-starter-stream-source-mail/pom.xml:
```diff
@@ -0,0 +1,65 @@
+		<dependency>
+			<groupId>javax.mail</groupId>
+			<artifactId>mail</artifactId>
+		</dependency>
```
Since we provide OOTB application for all know platforms and looks like no one is going to modify its classpath, we really should add an implementation for supported protocols:

``` xml
<dependency>
    <groupId>com.sun.mail</groupId>
    <artifactId>pop3</artifactId>
    <version>${sun-mail.version}</version>
</dependency>

<dependency>
    <groupId>com.sun.mail</groupId>
    <artifactId>imap</artifactId>
    <version>${sun-mail.version}</version>
</dependency>
```

BTW, I wonder how do those `MailSourceConfigurationTests` pass than?

artembilan replied:
Out of topic, of course, but I can't test it at all.
Maven can't find Gemfire artifact.
Maybe we missed something like: https://github.com/spring-cloud/spring-cloud-stream-modules/blob/master/.settings.xml#L34
?
A new GH issue ?

aamol replied:
Spring-integration-Java-Dsl uses "javax.mail:javax.mail-api" and "javax.mail:javax.mail-api" uses  "javax.mail:mail" so cant do much about it.

Also All the mail related Class like ImapStore are there in "javax.mail:mail"

artembilan replied:
Well, I have some clean up here locally. Will show you later as a polishing if there won't be so much concerns.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,107 @@
+/**
+ * A source module that listens for mail and emits the content as a message payload.
+ *
+ * @author Amol
```
The good rule to place here your full real name.
And I think your Git setting doesn't reflect your real name, too: `Amol Amol <aamol@sapient.com>` :smile: 

Yeah, minor, of course, but would be better to follow with Open Source process.

aamol replied:
My Name is Only "Amol" :)  . Company had a policy of mandatory second name hence Amol Amol

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2015 the original author or authors.
```
Minor, too.
But new classes should have a current year.

artembilan commented:
General comment. 
Shouldn't we support `ImapIdleChannelAdapter` as well? E.g. really via some property, like `idle=true/false`...

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,107 @@
+		}
+		else {
+			flowBuilder = IntegrationFlows.from(
+					Mail.pop3InboundAdapter(mailURL)
```
The `if...else` code could be simplified, when we extract `Mail...Adapter()` variable.
But yeah, minor.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+public class MailSourceProperties {
+
+	/**
+	 * Protocol to be used for recieving mail.
```
Typo: rec**e**i~~e~~ving

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+	private String password;
+
+	/**
+	 * Host URL for email access
```
Nit-pick, but it is just `host`. No any URL here.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+	private boolean delete = false;
+
+	/**
+	 * Comma seperated list of javaMail properties to be used.
```
Typo: sep**a**rated

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+		this.javaMailProperties = javaMailProperties;
+	}
+
+	protected Map<String, String> parsePropertiesString() {
```
I wonder if we could follow with existing approach in the `ScriptVariableProperties` which is applied onto the `spring-cloud-starter-stream-processor-groovy-transform` for example. I mean something like this:

```
/**
  * Variable bindings as a new line delimited string of name-value pairs, e.g. 'foo=bar\n baz=car'.
*/
private Properties variables;
```

Where we receive `Properties` as is and just inject them into the target adapter.
All other _hard_ parsing logic is done during properties population by Boot.

aamol replied:
ScriptVariableGeneratorConfiguration  gives a generic property "variable" so not very informative for java mail properties

Do you think I should replicate the behaviour for javaMailProperties ?

artembilan replied:
No, I just mean that `javaMailProperties` should be directly as a `Properties` object, without any manual conversions.
Similar to the `ScriptVariableProperties.variables` property.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/resources/META-INF/spring.provides:
```diff
@@ -0,0 +1 @@
+provides: spring-cloud-starter-stream-source-mail
```
New line in the end of each file.
The IDE has a configuration option on the matter.
Yes, it is minor, too, but that saves some time for all of us.
At least I won't write such a comment and you won't read :smile: 

Just some code style rule.

aamol replied:
Done :)

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/test/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfigurationTests.java:
```diff
@@ -0,0 +1,179 @@
+package org.springframework.cloud.stream.app.mail.source;
```
All classes must have `Copyright` header.

aamol replied:
Done

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/test/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfigurationTests.java:
```diff
@@ -0,0 +1,179 @@
+			"host=localhost", "folder=INBOX", "mark-as-read=true", "delete=false",
+			"java-mail-properties=mail.imap.socketFactory.fallback=true,mail.store.protocol=imap,mail.debug=true" })
+	public static class ImapPassTests extends MailSourceConfigurationTests {
+		private final static ImapServer imapServer = PoorMansMailServer.imap(50000);
```
The port really must be random:

``` java
private static PoorMansMailServer.MailServer MAIL_SERVER;

...
protected static void startMailServer(PoorMansMailServer.MailServer mailServer) throws InterruptedException {
    MAIL_SERVER = mailServer;
    System.setProperty("test.mail.server.port", "" + MAIL_SERVER.getPort());
    int n = 0;
    while (n++ < 100 && (!MAIL_SERVER.isListening())) {
        Thread.sleep(100);
    }
    assertTrue(n < 100);
}

@AfterClass
public static void cleanup() {
    System.clearProperty("test.mail.server.port");
    MAIL_SERVER.stop();
}


@IntegrationTest({ "protocol=imap", "port=${test.mail.server.port}", "username=user", "password=pw",
        "host=localhost", "folder=INBOX", "mark-as-read=true", "delete=false",
        "java-mail-properties=mail.imap.socketFactory.fallback=true,mail.store.protocol=imap,mail.debug=true" })
public static class ImapPassTests extends MailSourceConfigurationTests {

    @BeforeClass
    public static void startImapServer() throws Throwable {
        startMailServer(PoorMansMailServer.imap(0));
        }

    @Test
    public void testSimpleTest() throws Exception {
```

To avoid conflicts on the CI server.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/test/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfigurationTests.java:
```diff
@@ -0,0 +1,179 @@
+		public void testSimpleTest() throws Exception {
+
+			Message<?> received = messageCollector.forChannel(source.output()).poll(100,
+					TimeUnit.SECONDS);
```
M-m-m. 100 seconds seem for me so much to block the bad case on the CI server. 
10 would be enough for our mock mail server.

aamol replied:
Reduced sleep and poll time to 10 seconds

artembilan commented on pom.xml:
```diff
@@ -49,6 +49,7 @@
     <module>jms</module>
     <module>log</module>
     <module>metrics</module>
+	<module>mail</module>
```
@sobychacko , this is correct indentation -tabs. All the rest lines are with whitespaces.
Enough easy to fix via refactoring, but, please, reconsider your IDE options for XML formatting.

artembilan commented:
That's all from my side for now.
I can't say (yet) anything about the generated applications because I couldn't run it properly:
```
[ERROR] Failed to execute goal org.springframework.cloud.stream.app.plugin:spring-cloud-stream-app-maven-plugin:1.0.0.M1:generate-app (default-cli) on project spring-cloud-stream-app-generator: Execution default-cli of goal org.springframework.cloud.stream.app.p
lugin:spring-cloud-stream-app-maven-plugin:1.0.0.M1:generate-app failed: java.nio.file.DirectoryNotEmptyException: D:\SpringIO\spring-cloud\spring-cloud-stream-app-starters\apps\aggregate-counter-sink-kafka -> 
```
Should I?

aamol commented:
Thanks @artembilan 
I will go through these comments and update soon

Commit:
Encorporated comments from @artembilan

Commit:
Reduced poll time to 10 second

aamol commented:
@artembilan  I have fixed most of the comments. Please have a look and let me know if more changes are required.
I have re-factored MailSourceConfiguration to use ImapIdle adapter as well. Please have a look

Also  generate-app has a bug where it does not work on windows. I generated and tested from Linux

artembilan commented:
Hello @aamol !

Sorry for delay and thank you for update!
I'll come to you tomorrow for review but meanwhile would you mind to rebase your branch to the current `master`. 
We have just merged some clean over there in poms, so your one should become much simpler now.

Appreciate your patience and contribution.

aamol commented:
Done

Commit:
Minor Formatting changes and cleanup

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,239 @@
+	 * @return the protocol
+	 */
+	public String getProtocol() {
+		return protocol;
```
Minor, but our code style requires `this.` for properties.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+				break;
+			default:
+				adapterSpec = null;
+				break;
```
```
throw new IllegalArgumentException("Unsupported mail protocol: " + this.properties.getProtocol());
```

Instead.
No reason to go ahead if we don't understand the protocol.

aamol replied:
Agreed. Done with the change

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+		flowBuilder = getFlowBuilder();
+
+		return flowBuilder.transform(new MailToStringTransformer())
+				.channel(source.output()).get();
```
I found that this works too, but it is more cleaner:

``` java
return getFlowBuilder()
        .transform(Transformers.fromMail())
        .channel(Source.OUTPUT)
        .get();
```

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+		Properties javaMailProperties = new Properties();
+		String mailURL = properties.getProtocol() + "://" + properties.getUsername() + ":"
+				+ properties.getPassword() + "@" + properties.getHost() + ":"
+				+ properties.getPort() + "/" + properties.getFolder();
```
H-m-m. Having all these options as required, to be able to build the normal mail URL, I'd say that it would be better to request only a single `url` option, where I'm sure something like `port` or `login/password` maybe optional.
WDYT?

aamol replied:
I think taking each field separately is more intuitive to the user especially those who will be configuring mail in Spring Cloud DataFlow. I can add a field for complete  URL where either providing URL or providing variables individually can both work. Thoughts ?

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+
+		IntegrationFlowBuilder flowBuilder;
+		if (properties.isIdleImap()) {
+			flowBuilder = getIdleImapflow(mailURL);
```
`javaMailProperties` is missed for `idle` adapter.
With the new `Properties` style you can just get them from those `target adapter` methods directly from the `MailSourceProperties`:

```
this.properties.getJavaMailProperties()
```

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+	private IntegrationFlowBuilder getIdleImapflow(String mailURL) {
+		IntegrationFlowBuilder flowBuilder = null;
+		if ("imap".equalsIgnoreCase(properties.getProtocol())
+				|| "imaps".equalsIgnoreCase(properties.getProtocol())) {
```
I think this `if` does not make sense if we use `idleImap = true`.
The "wrong protocol" is a responsibility of the target `Mail.imapIdleAdapter()` already.

aamol replied:
Yes, Just checked `Mail.imapIdleAdapter()` . Removed `if` check

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+		if ("imap".equalsIgnoreCase(properties.getProtocol())
+				|| "imaps".equalsIgnoreCase(properties.getProtocol())) {
+			flowBuilder = IntegrationFlows
+					.from(Mail.imapIdleAdapter(mailURL).shouldReconnectAutomatically(true)
```
This `shouldReconnectAutomatically(true)` is redundant. It is like that by default. See `ImapIdleChannelAdapter.shouldReconnectAutomatically`

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/test/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfigurationTests.java:
```diff
@@ -0,0 +1,214 @@
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = MailSourceConfigurationTests.MailSourceApplication.class)
+@DirtiesContext
+@WebIntegrationTest(randomPort = true)
```
I found how to simplify your life instead:

``` java
@TestPropertySource(properties = { "port=${test.mail.server.port}", "username=user",
        "password=pw", "host=localhost", "folder=INBOX", "mark-as-read=true",
        "delete=false",
        "java-mail-properties=mail.imap.socketFactory.fallback=true\\n mail.store.protocol=imap\\n mail.debug=true" })
```

Where the config for the one of implementors looks like:

``` java
@IntegrationTest({ "protocol=imap", "idleImap=true" })
public static class ImapIdlePassTests extends MailSourceConfigurationTests {
```

I mean that I have moved all the common duplicated option to one place to avoid some mess.

aamol replied:
Thanks this makes it easy  . Will use this in my other test cases as well :)

artembilan commented on spring-cloud-stream-app-dependencies/pom.xml:
```diff
@@ -654,8 +660,11 @@
+			<dependency>
+				<groupId>javax.mail</groupId>
+				<artifactId>mail</artifactId>
+				<version>1.5.0-b01</version>
```
We don't need this dependency because Boot provides for us:

```
<dependency>
    <groupId>com.sun.mail</groupId>
    <artifactId>javax.mail</artifactId>
</dependency>
```

with the version of `1.5.5`.
That is the reason do not use `javax.mail:mail` anymore, because this one hasn't been updated since March 2013.

artembilan commented:
That's all with this review round.
Good progress!
Thank you, @aamol !

@sobychacko , do we have some documentation to update with such a new application which has to be address with the PR, too?

Commit:
Incorporated review comments

artembilan commented:
Just for heads up regarding the new feature in the [SI Mail module](http://docs.spring.io/spring-
integration/docs/4.3.0.RC1/reference/html/whats-new.html#_mail_message_mapping).

I just wonder if we can be more flexible ([Postal's Law](https://en.wikipedia.org/wiki/Robustness_principle) :smile:) and allow to convert the email content not only to string.

So, the question again: what is the release plans for this project? 
Can't we switch to to SI-4.3 and therefore to Boot-1.4 yet?
@sobychacko , @sabbyanandan

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,157 @@
+
+		flowBuilder = IntegrationFlows.from(Mail.imapIdleAdapter(mailURL)
+				.shouldDeleteMessages(this.properties.isDelete())
+				.javaMailProperties(this.properties.getJavaMailProperties())
```
This is redundant option, since you cover it in the main method.

artembilan replied:
Oh, No!
Never mind. It is a different branch for the `imapIdle` adapter.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,157 @@
+		String mailURL = this.properties.getProtocol() + "://"
+				+ this.properties.getUsername() + ":" + this.properties.getPassword()
+				+ "@" + this.properties.getHost() + ":" + this.properties.getPort() + "/"
+				+ this.properties.getFolder();
```
Well, my point about the only one full email option is that all these parts may be optional.
On the other hand it is a syntax of the [native](http://docs.spring.io/spring-integration/reference/html/mail.html#mail-inbound) SI adapters:

``` xml
<int-mail:imap-idle-channel-adapter id="customAdapter"
                    store-uri="imaps://foo.com:password@imap.foo.com/INBOX">

<mail:inbound-channel-adapter id="pop3Adapter"  store-uri="pop3:bar" >
```

But yeah... That may be just my humble opinion...

So, let's collect votes! :smile:

aamol replied:
Can change to URL but was more inclined towards Spring-XD format as I was using that before :smile:
Reference  : [Spring-XD Mail Configuration ](https://github.com/spring-projects/spring-xd/blob/master/modules/source/mail/config/mail.xml#L18-L20)

artembilan replied:
Yeah, yeah. I see that XD configuration, too.
And I can say that it is has the same vulnerabilities, when some of options aren't required and you miss it during definition.

I still insist that we should use only single `url` property with the sample in docs for proper syntax.
Although it may depend of the vendor like it is with JDBC. And I'm sure you know that we configure there only single URL:

```
spring.datasource.url= # JDBC url of the database.
```

No, any `host`/`port`, just one long url to meet the target server requirements.
Although I may agree that we can extract `login`/`password` as they can be pushed to the `javaMailProperties` and will be used properly from the target Mail Session:
http://connector.sourceforge.net/doc-files/Properties.html

aamol replied:
I see your point. Also as we are doing it in same format in other places makes more sense now to make it consistent.
I did the change.

artembilan commented on spring-cloud-stream-app-dependencies/pom.xml:
```diff
@@ -561,6 +568,11 @@
 			<dependency>
+				<groupId>javax.mail</groupId>
+				<artifactId>mail</artifactId>
+				<version>1.5.0-b01</version>
```
M-m-m. Not sure where my comment from here has gone.
But should not use `javax.mail` artifact.
Only `com.sun.mail:javax.mail`. 
And rely already on the Boot inheritance. 
Therefore remove this one.

aamol replied:
Replaced with `com.sun.mail:javax.mail`

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+	/**
+	 * JavaMail properties as a new line delimited string of name-value pairs, e.g. 'foo=bar\n baz=car'.
+	 */
+	private Properties javaMailProperties;
```
I'm afraid but this one can't be `null`.
Therefore some default to just `new Properties();`.

aamol replied:
Initialized with `new Properties();`

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+	/**
+	 * @return the javaMailProperties
+	 */
+	public Properties getJavaMailProperties() {
```
And `@NotNull` validation here according to the `non-null` requirements.

aamol replied:
We should not make it not-null as JavaMailProperties is an optional parameter. Can be if someone has a local SMTP server with minimal configuration and security

artembilan replied:
Right, it is optional and it is a `new Properties()` by default.
But it must be `not-null` anyway, since we can call `setJavaMailProperties(null);`
And it really can't be `null`.  See `Session` source code:

``` java
// Constructor is not public
    private Session(Properties props, Authenticator authenticator) {
    this.props = props;
    this.authenticator = authenticator;

    if (Boolean.valueOf(props.getProperty("mail.debug")).booleanValue())
        debug = true;
```

aamol replied:
Agreed. Added `@NotNull`

artembilan replied:
No need to comment if you agree.
Otherwise my iPhone beeps for me for a new email, but I see there only a "smap" :wink: 

For agreement there is just enough to add one "Pushed" comment after adding the commit with the fix.

Of course, I appreciate any disagreement. So, feel free to comment for that case with any way!

artembilan commented:
And I don't see any documentation changes, BTW.
@sobychacko , will be that fine to leave it as is and address the concern in the separate commit?
Thanks

Commit:
Comments Incorporation

- Added a test for Blank JavaMailProperties
- Replaced javax.mail with com.sun.mail:javax.mail 
- ReadMe for parameter details

aamol committed:
Rename README.md to README.adoc

aamol commented:
Hi @artembilan, Have updated code with the changes. Have seen the new SI code which removes the need for converting to String. Will try it out on local with new SI version.

And apologies for your effort of going through multiple rounds of review.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceProperties.java:
```diff
@@ -0,0 +1,220 @@
+ * @author Amol
+ */
+@ConfigurationProperties
+public class MailSourceProperties {
```
Looks you have missed some properties here since XD anyway:

```
private String expression = "true";
```

which should be use in the `ImapIdleChannelAdapterSpec` configuration:

```
/**
     * Configure a SpEL expression to select messages. The root object for the expression
     * evaluation is a {@link javax.mail.internet.MimeMessage} which should return a boolean
     * result (true means select the message).
     * @param selectorExpression the selectorExpression.
     * @return the spec.
     */
    public ImapIdleChannelAdapterSpec selectorExpression(String selectorExpression) {
        this.receiver.setSelectorExpression(PARSER.parseExpression(selectorExpression));
        return this;
    }
```

Plus let's consider the `IMAPS_PROPERTIES` anyway:

```
    private static final String IMAPS_PROPERTIES = "mail.imap.socketFactory.class=javax.net.ssl.SSLSocketFactory," +
            "mail.imap.socketFactory.fallback=false," +
            "mail.store.protocol=imaps";
```

Or does it work well with `imaps` and without these extra props?

Have you tested the solution against the real mail server BTW ?

aamol replied:
Added expression for selector. 
I think IMAP properties works fine with `javaMailProperties`. Same for POP3

Tested with gmail and exchange server for Imap and IdleImap 
Test on another server for POP3

Used the below parameters for IMAP . 
`--protocol=imaps --mail-url=imaps://username%40gmail.com:password@imap.gmail.com:993/INBOX --mark-as-read=true --delete=false --java-mail-properties=mail.imap.socketFactory.class=javax.net.ssl.SSLSocketFactory\nmail.imap.socketFactory.fallback=false\nmail.store.protocol=imaps\nmail.debug=true --spring.cloud.stream.bindings.output.destination=mail1`

sobychacko replied:
@artembilan we can always address the docs later, but a simple starter readme with the basic options available would be nice to have.

sobychacko replied:
Also, i think the plans are to release an RC early next month and then followed by GA. So, not sure about moving to SI 4.3, boot 1.4 etc. But @sabbyanandan can chime in here and clarify.

artembilan replied:
@sobychacko , thank you for your attention! 
I've seen some docs today already, but right now... @aamol , can you explain where is that `README` ?

Regarding `mail-url` and `protocol`. Looks weird. 
I really sure that we can extract protocol from the URL. Similar like `AbstractMailReceiver` does that:

```
public AbstractMailReceiver(String url) {
        if (url != null) {
            this.url = new URLName(url);
        }
        else {
            this.url = null;
        }
    }
```

and afterwards:

```
this.url.getProtocol();
```

artembilan replied:
I'd really move the `mail.imap.socketFactory.class` and similar to something internal, only `imaps` specific and merge that into `javaMailProperties`, if doesn't have yet.

aamol replied:
README updated at `mail/spring-cloud-starter-stream-source-mail/README.adoc`

Extracted common properties into a method and merged with `javaMailProperties`

Refactored to use URLName and its methods

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/test/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfigurationTests.java:
```diff
@@ -0,0 +1,229 @@
+
+	}
+
+}
```
Please, fix your IDE options to insert an new line in the end of files.

aamol replied:
Fixed now. Used the `eclipse-code-formatter.xml` mentioned in [https://github.com/spring-cloud/spring-cloud-starters](https://github.com/spring-cloud/spring-cloud-starters) . That had this issue.
We should fix that.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/README.adoc:
```diff
@@ -0,0 +1,27 @@
+//tag::ref-doc[]
+= Mail Source
```
Looks like this this doc must be mentioned in the `spring-cloud-stream-app-starters-docs/../sources.adoc`

aamol replied:
Done

artembilan commented:
BTW, @aamol , we are close to make it working to generate `apps` on Windows: https://github.com/spring-cloud/spring-cloud-stream-app-maven-plugin/pull/5 !

Commit:
Updated code for review comments

- Updated Sources.adoc for Readme
-Update code to use mailURL instead of individual parameters

aamol commented:
Hi @artembilan , Made the changes. 

Saw the maven plugin fix. Works fine on my machine :smile:

artembilan commented:
> .It now works fine on my machine

That's great! 
Would you mind adding kinda `.bat` script for apps generation to this your PR ?
Thank you!

artembilan commented:
> to release an RC early next month and then followed by GA.

That is fully with SI-4.3.0.RELEASE early the next month, too. Just after SF-4.3.0.RELEASE. 
So, if we switch now and fix all the incompatibility, we will be fine with your release plans.

But yes... Looking closely to what we have here with the Maven dependency management, that won't be so easy to keep Boot 1.3 for parent Spring Cloud and use Boot 1.4 in this project.
I'm newbie in this "train" process, but I feel that  for consistency we can't move so fast with dependencies.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/test/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfigurationTests.java:
```diff
@@ -0,0 +1,207 @@
+
+	}
+
+	@IntegrationTest({ "protocol=imap", "idleImap=true","mail-url=imap://user:pw@localhost:${test.mail.server.port}/INBOX" })
```
M-m-m. Does it really work for properties camel sty and dash delimiter? Cool! :smile:

artembilan commented:
@aamol , I think I'm fine with what we have her already. 
I can fix all those minor concerns on merge, but would like hearing from your first of all.
Especially about that not related `.bat` scrip :smile: 

Very valuable contribution and sane discussion!
Thank you very much and looking forward for more!

Commit:
Code updated for default properties

- Added default propeties for POP3,POP3S,IMAP,IMAPS
- Added generateApps.bat
- Updated README
- Updated Tests
- Refactored protocol extraction

aamol commented:
Hi @artembilan , 

Was a great experience contributing to Spring. Have learnt a lot in the process around the general coding standards and guideline followed in Spring community.

Looking forward to some more items where I can contribute.
Thanks a lot.

P.S. I have added bat file as well as some other re-factoring suggested by you..

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,201 @@
+		}
+		else {
+			urlName = null;
+		}
```
This `if...else` logic around `null` is redundant. You already said that `mailUrl` that is it `@NotNull`.
On the other hand the `@ConfigurationProperties` is enough magic with the `ConversionService` behind and we can just have it like:

```
private URLName mailUrl;
...
@NotNull
public URLName getMailUrl() {
    return mailUrl;
}

...
public void setMailUrl(URLName mailUrl) {
    this.mailUrl = mailUrl;
}
```

in the `MailSourceProperties`. And don't worry about this string conversions manually.

In other words: that simplifies our life! :smile:

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,201 @@
+	Source source;
+
+	@Autowired
+	MailSourceProperties properties;
```
Minor right now and I'll fix on merge, but for the future.
Properties must be `private`, just for better code style and other unexpected side-effects.

artembilan commented on mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java:
```diff
@@ -0,0 +1,201 @@
+			case "IMAPS":
+				adapterSpec = getImapFlowBuilder(urlName);
+				break;
+			case "POP3":
```
The `"POP3S":` must be here as well.
Will fix on merge

artembilan commented:
Here is the polishing for confirmation: https://github.com/artembilan/spring-cloud-stream-app-starters/commit/63fed0ba1d0661b60096c8a731911d08eacda53a.

Note: I was forced to change the command in the `generateApps.bat` to use Maven Wrapper (`mvnw`). Just because with the local Maven I end up with:
```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:exec (imports) on project spring-cloud-stream-app-generator: Command execution failed. Cannot run program "egrep" (in directory "D:\SpringIO\spring-cloud\spring-cloud-stream-app-starters\spring-cloud-stream-app-generator"): CreateProcess error=2, ?? ??????? ????? ????????? ???? -> [Help 1]
```
Maybe that is a result of the Maven version incompatibility (my one is `3.1.1`), but from other side what is the point to rely on the local one, if we always are good to make end-user to use that one which comes together with the project? The same we have with Gradle and everybody is happy. Isn't it? :smile: 
CC @sobychacko

sobychacko commented:
@artembilan @aamol Looks good to me

artembilan commented:
Thank you @sobychacko !
So, merging finally ! :smile:

artembilan commented:
Merged as https://github.com/spring-cloud/spring-cloud-stream-app-starters/commit/d95aff1fb10639c7687d89e0300df943818385eb.

Thank you very much, @aamol , for the contribution!
That was good experience for me as well to work in this project and discussion with you.
So, looking for more!

I even see the next your steps #53 ?

@sobychacko , do we need to register this new app somehow for SCDF, like we did for modules before? What is the algorithm now ?

sabbyanandan commented:
@aamol - Thank you for the contribution! 😄

sobychacko commented:
@artembilan For dataflow-server-local, you can register this through here: https://github.com/spring-cloud/spring-cloud-dataflow/tree/master/spring-cloud-dataflow-server-local/src/main/resources/META-INF.

For other dataflow variants, you have to manually register the modules. I think there is an issue in progress where they would read from the descriptors provided by the app-starters.

artembilan commented:
So, Do I ( @aamol ? ) have to raise a PR against SCDF to let `source.mail` to appear in those properties ?
